library/pkcs11_client.c

```diff
@@ -0,0 +1,527 @@
+    if( rv != CKR_OK )
+        goto exit;
+
+    if( mechanism.mechanism == CKM_ECDSA )
+    {
+        /* The signature from the token contains r and s concatenated,
+         * each in the form of a big-endian byte sequence, with r and s
+         * having the same length as the base point.
+         *
+         * A standard ECDSA signature is encoded in ASN.1:
+         *   SEQUENCE {
+         *     r INTEGER,
+         *     s INTEGER
+         *   }
+         *
+         * Perform the conversion using existing utility functions,
+         * with temporary bignums.
+         */
+        uint16_t byte_len = ( ( ctx->bit_length + 7 ) / 8 );
+        size_t sig_size = MBEDTLS_ECDSA_MAX_SIG_LEN( ctx->bit_length );
+        mbedtls_mpi r, s;
+        mbedtls_mpi_init( &r );
+        mbedtls_mpi_init( &s );
+        rv = CKR_OK;
+        if( ck_sig_len != 2 * byte_len )
+        {
+            /* Bad data from the token */
+            rv = CKR_GENERAL_ERROR;
+            goto ecdsa_exit;
+        }
+        if( mbedtls_mpi_read_binary( &r, sig, byte_len ) != 0 ||
+            mbedtls_mpi_read_binary( &s, sig + byte_len, byte_len ) != 0 )
+        {
+            rv = CKR_HOST_MEMORY;
+            goto ecdsa_exit;
+        }
+        /* The signature buffer is guaranteed to have enough room for
+           the encoded signature by the pk_sign interface. */
+        if( ecdsa_signature_to_asn1( &r, &s, sig, sig_len, sig_size ) != 0 )
+        {
+            rv = CKR_GENERAL_ERROR;
+            goto ecdsa_exit;
+        }
+    ecdsa_exit:
+        mbedtls_mpi_free( &r );
+        mbedtls_mpi_free( &s );
+        if( rv != CKR_OK )
+            goto exit;
+    }
+    else
+    {
+        *sig_len = ck_sig_len;
+    }
+
+exit:
+    if( rv != CKR_OK )
+        memset( sig, 0, ck_sig_len );
+    return( pkcs11_err_to_mbedtls_pk_err( rv ) );
+}
+
+static const mbedtls_pk_info_t mbedtls_pk_pkcs11_info =
+    MBEDTLS_PK_OPAQUE_INFO_1( "pkcs11"
+                              , pkcs11_pk_get_bitlen
+                              , pkcs11_pk_can_do //can_do
+                              , pkcs11_pk_signature_size
+                              , NULL //pkcs11_verify
+                              , pkcs11_sign
+                              , NULL //pkcs11_decrypt
+                              , NULL //pkcs11_encrypt
+                              , NULL //check_pair_func
+                              , pkcs11_pk_alloc
+                              , pkcs11_pk_free
+                              , NULL //debug_func
+        );
+
+int mbedtls_pk_setup_pkcs11( mbedtls_pk_context *ctx,
+                             CK_SESSION_HANDLE hSession,
+                             CK_OBJECT_HANDLE hPublicKey,
+                             CK_OBJECT_HANDLE hPrivateKey )
+{
+    CK_OBJECT_CLASS public_key_class = -1, private_key_class = -1;
+    CK_KEY_TYPE public_key_type = -1, private_key_type = -1;
+    mbedtls_pk_type_t can_do;
```

Comment:
It's what `can_do` will say yes to, that's where the name comes from. It isn't a `repr_type` because it isn't about how the key is represented: the representation is inside the crypto token and we have no way to know what it is. This is the semantic type of the key.

Reply:
Indeed, `repr_type` would not be good. Let's keep with `can_do` then.